### PR TITLE
ENH: peer forwarder domain or ip validator

### PR DESCRIPTION
**Pull request by chenqi0805**
*Issue #, if available:*
#437 

*Description of changes:*

-----
This PR adds validation before initializing a peer provider. 

Note: I have verified that invalid endpoint in addn-hosts of dnsmasq, e.g.

```
10.10.1.2 prepper-cluster
10_10.1.3 prepper-cluster
```

will not be ack by endpointGroup:

```
2021-04-20T16:33:17,543 [main] INFO  com.amazon.dataprepper.plugins.prepper.peerforwarder.discovery.DnsPeerListProvider - Found endpoints: [Endpoint{prepper-cluster, ipAddr=10.10.1.2, weight=1000}]
```
while valid endpoint:
```
10.10.1.2 prepper-cluster
10.10.1.3 prepper-cluster
```
will still be Ack:
```
2021-04-20T16:30:23,453 [main] INFO  com.amazon.dataprepper.plugins.prepper.peerforwarder.discovery.DnsPeerListProvider - Found endpoints: [Endpoint{prepper-cluster, ipAddr=10.10.1.2, weight=1000}, Endpoint{prepper-cluster, ipAddr=10.10.1.3, weight=1000}]

```
so there is no need to do validity check in `getPeerList()`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opendistro-for-elasticsearch/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/PeerListProviderFactory.java [hunk 1]
```diff
@@ -36,10 +40,15 @@ public PeerListProvider createProvider(final PluginSetting pluginSetting) {
             case STATIC:
                 final List<String> endpoints = (List<String>) pluginSetting.getAttributeOrDefault(PeerForwarderConfig.STATIC_ENDPOINTS, null);
                 Objects.requireNonNull(endpoints, String.format("Missing '%s' configuration value", PeerForwarderConfig.STATIC_ENDPOINTS));
+                Preconditions.checkArgument(endpoints.stream().allMatch(this::validateEndpoint), "Including invalid endpoints: %s", endpoints);
```
Comment (wrijeff) [inline data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/PeerListProviderFactory.java:43]: This won't indicate which endpoints are actually invalid. 

Might be better to stream-create a list of invalid endpoints, assert the list is empty, and if not then display the list contents to the user.

File: data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/PeerListProviderFactory.java [hunk 2]
```diff
@@ -3,7 +3,10 @@
 import com.amazon.dataprepper.metrics.PluginMetrics;
 import com.amazon.dataprepper.model.configuration.PluginSetting;
 import com.amazon.dataprepper.plugins.prepper.peerforwarder.PeerForwarderConfig;
+import com.google.common.base.Preconditions;
 import com.linecorp.armeria.client.endpoint.dns.DnsAddressEndpointGroup;
+import org.apache.commons.validator.routines.DomainValidator;
+import org.apache.commons.validator.routines.InetAddressValidator;
```
Comment (wrijeff) [inline data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/PeerListProviderFactory.java:9]: 👍

File: data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/PeerListProviderFactory.java [hunk 3]
```diff
@@ -36,10 +41,16 @@ public PeerListProvider createProvider(final PluginSetting pluginSetting) {
             case STATIC:
                 final List<String> endpoints = (List<String>) pluginSetting.getAttributeOrDefault(PeerForwarderConfig.STATIC_ENDPOINTS, null);
                 Objects.requireNonNull(endpoints, String.format("Missing '%s' configuration value", PeerForwarderConfig.STATIC_ENDPOINTS));
+                final List<String> invalidEndpoints = endpoints.stream().filter(endpoint -> !this.validateEndpoint(endpoint)).collect(Collectors.toList());
+                Preconditions.checkArgument(invalidEndpoints.size() == 0, "Including invalid endpoints: %s", invalidEndpoints);
```
Comment (wrijeff) [inline data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/PeerListProviderFactory.java:45]: nit - Prefer `isEmpty()` here

Comment (wrijeff) [reply data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/PeerListProviderFactory.java:45]: Also considering the use of [checkArgument](https://guava.dev/releases/19.0/api/docs/com/google/common/base/Preconditions.html#checkArgument(boolean,%20java.lang.Object)) vs [checkState](https://guava.dev/releases/19.0/api/docs/com/google/common/base/Preconditions.html#checkState(boolean,%20java.lang.Object))... The list of endpoints isn't exactly an argument being passed to the method, but it _is_ sort of passed in via the config file.

There's also the Apache Commons [Validate method](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/Validate.html#isTrue-boolean-), so many choices :P

**Review by wrijeff — APPROVED** [review]

File: data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/PeerListProviderFactory.java (hunk 3 above)
Comment (chenqi0805) [reply data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/PeerListProviderFactory.java:45]: makes sense to use checkState

**Review by wrijeff — APPROVED** [review]
👍

**Pull request merged**